Add ERC: Permissioned Pull

recurmj opened the pull request:
This PR proposes two related ERCs that together define a portable, revocable pull-permission primitive:

1. **Permissioned Authorization Object (PPO)**  
   - EIP-712 typed struct describing bounded, revocable consent to pull tokens.  
   - Chain-agnostic, portable across executors that adopt the same type.  
   - Intended as the shared “authorization primitive” layer.

2. **Permissioned Pull Execution Interface**  
   - Standard pull executor interface that consumes a PPO and executes `transferFrom` under strict rules.  
   - Defines `pull()`, nonce & revocation views, canonical events, and normative execution behavior.

Both drafts are aligned with the architectural analysis in [RIP-000 / “The One-Shot Authorization Flaw in Digital Value Systems”](https://ethereum-magicians.org/t/recur-rip-000-the-one-shot-authorization-flaw-in-digital-value-systems/27042) and the updated RIP-001 thread. Philosophical framing has been removed; these drafts focus purely on:

- defining a neutral, reusable EIP-712 Authorization object (PPO), and  
- a minimal, composable execution interface that can be implemented by any executor.

The intention is for these to serve as the canonical ERC-A / ERC-B split that James and others suggested earlier:

- **ERC-A**: Authorization object (PPO)  
- **ERC-B**: Pull execution interface

I’m happy to adapt naming, file layout, or minor spec details to fit current ERC process preferences.

recurmj committed:
Create erc-ppo-authorization-object.md

Add ERC-A: Permissioned Authorization Object (PPO)

recurmj committed:
Create erc-permissioned-pull-executor.md

Add ERC-B: Permissioned Pull Execution Interface

abcoathup commented on ERCS/erc-ppo-authorization-object.md:
```diff
@@ -0,0 +1,217 @@
+---
+eip: <to be assigned>
```
```suggestion
eip: 8103
```
Assigning next sequential EIP/ERC/RIP number.
Numbers are assigned by editors & associates.

Please also update the filename.

recurmj committed:
Update and rename erc-ppo-authorization-object.md to erc-8103.md

recurmj committed:
Update and rename erc-permissioned-pull-executor.md to erc-8102.md

recurmj committed:
Rename erc-8103.md to erc-8103.md

recurmj commented:
Thanks for the guidance.

I’ve split the specification into ERC-8102 (execution) and ERC-8103 (authorization), updated filenames, cross-references, titles, and discussions-to links accordingly.

Let me know if anything else is needed.

recurmj committed:
Update erc-8102.md

Removed the word “standard” in description

abcoathup commented on ERCS/erc-8103.md:
```diff
@@ -0,0 +1,217 @@
+---
+eip: <8103>
```
```suggestion
eip: 8103
```

abcoathup commented on ERCS/erc-8103.md:
```diff
@@ -0,0 +1,217 @@
+title: Permissioned Authorization Object (PPO)
+description: "A portable, revocable EIP-712 authorization struct for bounded pull-based transfers."
+author: "Mats Heming Julner (@recurmj)"
+discussions-to: "https://ethereum-magicians.org/t/erc-permissioned-authorization-object/27135"
```
```suggestion
discussions-to: https://ethereum-magicians.org/t/erc-8103-permissioned-authorization-object/27135
```

recurmj committed:
Update erc-8103.md

recurmj committed:
Update erc-8103.md

Added EIP links

recurmj committed:
Update erc-8103.md

recurmj committed:
Update erc-8103.md

recurmj committed:
Update erc-8102.md

Add EIP links

recurmj committed:
Update erc-8102.md

recurmj committed:
Update erc-8102.md

Fix license link

recurmj committed:
Update erc-8103.md

Fix license link

recurmj committed:
Update erc-8102.md

recurmj committed:
Update erc-8103.md

recurmj committed:
Update erc-8102.md

recurmj committed:
Update erc-8103.md

Fix typo malleability

recurmj committed:
Update erc-8102.md

Added a security consideration line and fixed 8103 link to be consistent

recurmj committed:
Update erc-8103.md

Fixed 8103/8102 reference links for consistency

SamWilsn commented on ERCS/erc-8102.md:
```diff
@@ -0,0 +1,217 @@
+- the `pull` function signature,
+- required view functions for nonce and domain separation,
+- canonical events and errors,
+- normative execution rules (validation, revocation, transfer).
```
This does a good job covering what your proposal is, but doesn't really include the purpose. Doesn't have to be long (that's what the motivation is for), but a sentence or two about why we need this proposal is usually good in the abstract.

SamWilsn commented on ERCS/erc-8102.md:
```diff
@@ -0,0 +1,217 @@
+- Implementations MUST consider token quirks (fee-on-transfer, non-standard return values) and SHOULD document behavior with such tokens.
+- Time windows reduce exposure of leaked signatures; however, key security is still critical.
+- Malicious grantees cannot exceed `maxPerPull` in a single call, but they MAY call `pull` multiple times if the authorization semantic allows. Systems requiring cumulative caps SHOULD implement them in higher layers or registries.
+- Implementations MUST reject malleable signatures by enforcing [EIP-2](./eip-2.md) rules (low-`s` values and valid `v`), as specified in ERC-8103.
```
Requirements necessary to make a secure implementation should be included in the specification section, but do discuss their implications here.

recurmj committed:
Update erc-8102.md

recurmj committed:
Update erc-8103.md

SamWilsn commented on ERCS/erc-8103.md:
```diff
@@ -0,0 +1,222 @@
+
+This EIP intentionally separates:
+
+- the **authorization object** (ERC-8103), and
```
```suggestion
- the **authorization object** (defined herein), and
```

SamWilsn commented on ERCS/erc-8103.md:
```diff
@@ -0,0 +1,222 @@
+
+## Rationale
+
+- **Separation of concerns**: ERC-8103 defines “what is consent?” while ERC-8102 defines “how is consent executed?”. This keeps wallets, AA stacks, and registries decoupled from any specific executor implementation.
```
```suggestion
- **Separation of concerns**: This proposal defines “what is consent?” while ERC-8102 defines “how is consent executed?”. This keeps wallets, AA stacks, and registries decoupled from any specific executor implementation.
```

SamWilsn commented on ERCS/erc-8103.md:
```diff
@@ -0,0 +1,222 @@
+Executors verifying PPOs MUST:
+
+- enforce validity windows (`validAfter`, `validBefore`) before execution,
+- enforce EIP-2 signature constraints (low-`s`, valid `v`),
```
```suggestion
- enforce [EIP-2](./eip-2.md) signature constraints (low-`s`, valid `v`),
```

SamWilsn commented on ERCS/erc-8102.md:
```diff
@@ -0,0 +1,221 @@
+
+The rules required for secure behavior are specified normatively in the Specification section. This section discusses implications and common pitfalls:
+
+- Implementations MUST consider token quirks (fee-on-transfer, non-standard return values) and SHOULD document behavior with such tokens.
```
```suggestion
- Implementations must consider token quirks (fee-on-transfer, non-standard return values) and should document behavior with such tokens.
```

SamWilsn commented on ERCS/erc-8102.md:
```diff
@@ -0,0 +1,221 @@
+
+- Implementations MUST consider token quirks (fee-on-transfer, non-standard return values) and SHOULD document behavior with such tokens.
+- Time windows reduce exposure of leaked signatures; however, key security is still critical.
+- Malicious grantees cannot exceed `maxPerPull` in a single call, but they MAY call `pull` multiple times if the authorization semantic allows. Systems requiring cumulative caps SHOULD implement them in higher layers or registries.
```
```suggestion
- Malicious grantees cannot exceed `maxPerPull` in a single call, but they can call `pull` multiple times if the authorization semantic allows. Systems requiring cumulative caps should implement them in higher layers or registries.
```

SamWilsn committed:
Apply suggestions from code review

Co-authored-by: Sam Wilson <57262657+SamWilsn@users.noreply.github.com>
